rust/vectordb/src/table.rs

```diff
@@ -227,46 +327,22 @@ impl Table {
                     message: e.to_string(),
                 },
             })?;
-        Ok(Table {
+        Ok(NativeTable {
             name: name.to_string(),
             uri: uri.to_string(),
-            dataset: Arc::new(dataset),
+            dataset: Arc::new(Mutex::new(dataset)),
             store_wrapper: write_store_wrapper,
         })
     }
 
-    /// Schema of this Table.
-    pub fn schema(&self) -> SchemaRef {
-        Arc::new(self.dataset.schema().into())
-    }
-
     /// Version of this Table
     pub fn version(&self) -> u64 {
-        self.dataset.version().version
-    }
-
-    /// Create index on the table.
-    pub async fn create_index(&mut self, index_builder: &impl VectorIndexBuilder) -> Result<()> {
-        let mut dataset = self.dataset.as_ref().clone();
-        dataset
-            .create_index(
-                &[index_builder
-                    .get_column()
-                    .unwrap_or(VECTOR_COLUMN_NAME.to_string())
-                    .as_str()],
-                IndexType::Vector,
-                index_builder.get_index_name(),
-                &index_builder.build(),
-                index_builder.get_replace(),
-            )
-            .await?;
-        self.dataset = Arc::new(dataset);
-        Ok(())
+        self.dataset.lock().expect("lock poison").version().version
     }
 
     /// Create a scalar index on the table
-    pub async fn create_scalar_index(&mut self, column: &str, replace: bool) -> Result<()> {
-        let mut dataset = self.dataset.as_ref().clone();
+    pub async fn create_scalar_index(&self, column: &str, replace: bool) -> Result<()> {
+        let mut dataset = self.clone_inner_dataset()?;
```

Comment:
I'd expect operations like this to reset the inner dataset.  Does this work this way for backwards compatibility?

Reply:
Lemme reset the dataset then. It was a left over from the previous behavior.